src/mallocTracer.cpp

```diff
@@ -0,0 +1,230 @@
+    MallocTracer::recordFree(addr, OS::nanotime());
+}
+
+#ifdef __linux__
+extern "C" WEAK DLLEXPORT void free(void* addr) {
+    if (unlikely(!_orig_free)) {
+        return;
+    }
+
+    if (likely(MallocTracer::initialized())) {
+        return free_hook(addr);
+    }
+    return _orig_free(addr);
+}
+#endif // __linux__
+
+u64 MallocTracer::_interval;
+volatile u64 MallocTracer::_allocated_bytes;
+
+Mutex MallocTracer::_patch_lock;
+int MallocTracer::_patched_libs = 0;
+bool MallocTracer::_initialized = false;
+
+__attribute__((constructor)) static void getOrigAddresses() {
+    // Store these addresses, regardless of MallocTracer being enabled or not.
+    _orig_calloc = ADDRESS_OF(calloc);
+    _orig_free = ADDRESS_OF(free);
+    _orig_malloc = ADDRESS_OF(malloc);
+    _orig_realloc = ADDRESS_OF(realloc);
+}
+
+bool MallocTracer::initialize() {
+    if (!__sync_bool_compare_and_swap(&_initialized, false, true)) {
+        return false;
+    }
+
+    CodeCache* lib = Profiler::instance()->findLibraryByName("libasyncProfiler");
+    assert(lib);
+
+    lib->mark(
+        [](const char* s) -> bool {
+            return strncmp(s, "_ZL11malloc_hook", 16) == 0 || strncmp(s, "_ZL11calloc_hook", 16) == 0 || strncmp(s, "_ZL12realloc_hook", 16) == 0 || strncmp(s, " _ZL9free_hook", 13) == 0;
+        },
+        MARK_ASYNC_PROFILER);
+
+    return installHooks();
+}
+
+bool MallocTracer::patchLibs(bool install) {
+    if (!initialized()) {
+        return false;
+    }
+
+    MutexLocker ml(_patch_lock);
+    if (!install) {
+        assert(_orig_malloc);
+        assert(_orig_calloc);
+        assert(_orig_realloc);
+        assert(_orig_free);
+
+        _patched_libs = 0;
+    }
+
+    CodeCacheArray* native_libs = Profiler::instance()->nativeLibs();
+    int native_lib_count = native_libs->count();
+
+    while (_patched_libs < native_lib_count) {
+        CodeCache* cc = (*native_libs)[_patched_libs++];
+
+        cc->patchImport(im_malloc, (void*)(install ? malloc_hook : _orig_malloc));
+        cc->patchImport(im_calloc, (void*)(install ? calloc_hook : _orig_calloc));
+        cc->patchImport(im_realloc, (void*)(install ? realloc_hook : _orig_realloc));
+        cc->patchImport(im_free, (void*)(install ? free_hook : _orig_free));
+    }
+
+    if (!install) {
+        _patched_libs = 0;
+    }
+
+    return true;
+}
+
+void MallocTracer::recordMalloc(void* address, size_t size, u64 time) {
+    if (updateCounter(_allocated_bytes, size, _interval)) {
+        MallocEvent event;
+        event._start_time = time;
+        event._address = (uintptr_t)address;
+        event._size = size;
+
+        Profiler::instance()->recordSample(NULL, size, MALLOC_SAMPLE, &event);
+    }
+}
+
+void MallocTracer::recordFree(void* address, u64 time) {
+    MallocEvent event;
+    event._start_time = time;
+    event._address = (uintptr_t)address;
+    event._size = 0;
+
+    Profiler::instance()->recordSample(NULL, 0, MALLOC_SAMPLE, &event);
```

Comment:
`free` calls are not sampled according to `_interval`.
It may be too expensive to record stack traces for each `free`. I therefore propose to record `free` events without a stack trace.